test/extended/apiserver/latency.go

```diff
@@ -0,0 +1,84 @@
+package apiserver
+
+import (
+	"context"
+	"fmt"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+	promhelper "github.com/openshift/origin/test/extended/util/prometheus"
+)
+
+var (
+	oc                   *exutil.CLI
+	url, bearerToken, ns string
+	execPod              *v1.Pod
+)
+
+var _ = g.Describe("[sig-instrumentation][Late] API Server latency", func() {
+	defer g.GinkgoRecover()
+
+	oc = exutil.NewCLIWithoutNamespace("apiserver-latency")
+	ns = oc.SetupNamespace()
+
+	// TODO: access Prometheus via route instead of exec pod
+	execPod = exutil.CreateExecPodOrFail(oc.AdminKubeClient(), ns, "execpod")
+	defer func() {
+		oc.AdminKubeClient().CoreV1().Pods(ns).Delete(context.Background(), execPod.Name, *metav1.NewDeleteOptions(1))
+	}()
+
+	g.BeforeEach(func() {
+		err := exutil.WaitForAnImageStream(
+			oc.AdminImageClient().ImageV1().ImageStreams("openshift"), "tools",
+			exutil.CheckImageStreamLatestTagPopulated, exutil.CheckImageStreamTagNotFound,
+		)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		var ok bool
+		url, _, bearerToken, ok = promhelper.LocatePrometheus(oc)
+		if !ok {
+			e2e.Failf("Prometheus could not be located on this cluster, failing prometheus test")
+		}
+	})
+
+	g.It("should verify that the apiserver requests latency are within expected thresholds", func() {
+		// 99th percentile of resource-scoped requests latency
+		scope := "resource|"
+		expectRequestLatency("GET", scope, 0.075)
+		expectRequestLatency("POST", scope, 0.2)
+		expectRequestLatency("PUT", scope, 0.075)
+		expectRequestLatency("PATCH", scope, 0.075)
+		expectRequestLatency("DELETE", scope, 0.075)
+
+		// 99th percentile of namespace-scoped requests latency
+		scope = "namespace"
+		expectRequestLatency("LIST", scope, 0.25)
+		expectRequestLatency("POST", scope, 0.75)
+		expectRequestLatency("DELETE", scope, 0.75)
+
+		// 99th percentile of cluster-scoped requests latency
+		scope = "cluster"
+		expectRequestLatency("LIST", scope, 1)
+		expectRequestLatency("POST", scope, 0.5)
+	})
+})
+
+func expectRequestLatency(verb string, scope string, threshold float64) {
+	resp, err := promhelper.RunQuery(
+		fmt.Sprintf(
+			`histogram_quantile(0.99, sum(rate(apiserver_request_duration_seconds_bucket{job="apiserver",verb="GET",scope=~"%s"}[%s])) by (le))`,
+			scope, exutil.DurationSinceStartInSeconds().String()
+		),
+		ns, execPod.Name, url, bearerToken,
+	)
+
+	o.Expect(err).NotTo(o.HaveOccurred())
+	o.Expect(resp.Data.Result).NotTo(o.BeEmpty())
+	o.Expect(resp.Data.Result[0].Value).Should(o.BeNumerically("<"), threshold)
```

Comment:
we want to control this message very exactly so that it indicates somethign like

"p99 of requests like this (for example: list of configmaps across all namespaces) have a historical p99 of duration.  We set the threshold to other-duration.  This job run has a p99 of third-duration"

Reply:
I changed it to:
```go
	o.Expect(latency).Should(o.BeNumerically("<"), threshold,
		fmt.Sprintf("expected 99th percentile of apiserver request latency with verb=%q and scope=%q to be inferior to %s, got %s", verb, scope, threshold, latency),
	)
```
Doing it per resource would be very tedious here. To me, this should be a very high-level warning that something is degraded and further investigation is needed to figure exactly what.
Since the investigation is far from being trivial as we've seen during the past couple of days, it might be worth attaching a link explaining the steps that need to be taken in order to figure out what is wrong, similar to the document I created. wdyt?